Allows internal representation of DomainMatrix to be both sparse and dense

ks147 opened the pull request:
#### References to other Issues or PRs
This completes #21140 


#### Brief description of what is fixed or changed
The DomainMatrix class was originally introduces with DDM as its
internal representation representing a dense internal format. The SDM
sparse format was added later and made to be the internal format of DDM
always. This commit makes it possible DomainMatrix to use either SDM or
DDM as its internal format in the constructor by passing fmt="dense" or
fmt="sparse".

#### Other comments
Also added functions `to_sparse` and `to_dense`
This PR also implements binary operations between sparse 
and dense matrices in DomainMatrix class

#### Release Notes

<!-- BEGIN RELEASE NOTES -->
* polys
  * Added internal representation as both dense and sparse matrix
<!-- END RELEASE NOTES -->

ks147 committed:
Allow DomainMatrix to to be dense or sparse

The DomainMatrix class was originally introduces with DDM as its
internal representation representing a dense internal format. The SDM
sparse format was added later and made to be the internal format of DDM
always. This commit makes it possible DomainMatrix to use either SDM or
DDM as its internal format in the constructor by passing fmt="dense" or
fmt="sparse".

Co-authored-by: Kartik Sethi <kartiks31416@gmail.com>
Co-authored-by: Oscar Benjamin <oscar.j.benjamin@gmail.com>

ks147 committed:
Added functions to_dense and to_sparse

ks147 committed:
Added Tests

ks147 committed:
Added Binary operations between sparse and dense matrix

ks147 committed:
Fixed examples in domainmatrix.py

ks147 committed:
Added hstack function to DDM class

ks147 commented:
@oscarbenjamin, please review. I've also added you as the co-author.

oscarbenjamin commented on sympy/polys/matrices/domainmatrix.py:
```diff
@@ -236,6 +310,56 @@ def to_field(self):
+        >>> from sympy import QQ
+        >>> A = DomainMatrix([[1, 0],[0, 2]], (2, 2), QQ)
+        >>> A
+        DomainMatrix([[1, 0], [0, 2]], (2, 2), QQ)
```
I guess this should be `A.rep`

oscarbenjamin commented on sympy/polys/matrices/domainmatrix.py:
```diff
@@ -357,6 +478,11 @@ def hstack(A, B):
 
         """
+
+        if isinstance(A.rep, list) != isinstance(B.rep, list):
+            A = A.to_dense()
+            B = B.to_dense()
+
         A, B = A.unify(B)
```
The logic for combining dense and sparse could be handled in `unify`. Perhaps there could be an option to specify whether to unify towards sparse or dense if they are mixed or whether to always convert to dense or sparse or something like that. I think in basic terms `A.unify(B)` should make the two matrices compatible for subsequent operations.

oscarbenjamin replied:
There could also be separate functions like `A.unify_domain` and `A.unify_fmt` and then general `A.unify` could call both.

oscarbenjamin replied:
Maybe operations like `a + b` should be changed to unify the domains rather than raising an exception.

ks147 replied:
Ok, I'll make those changes. 
What do you think of converting both matrices to dense if their `rep` doesn't match. Should there be an option to convert both to sparse?

oscarbenjamin replied:
I'm not entirely sure whether everything should unify automatically or whether explicit conversion should be done by the user of `DomainMatrix`. Still thinking about it...

For now to finish this you could just make it so that mixing dense and sparse is an error. Then perhaps later if we want automatic conversions we can add them.

ks147 replied:
Well I did unify automatically, for binary operations in latest commit. I saw this message after pushing that change

oscarbenjamin commented:
There should always be tests added for any new functionality. For example if it is now possible to add a dense and a sparse DomainMatrix then there needs to be a test for that. You should also check code coverage of the tests.

ks147 commented:
> There should always be tests added for any new functionality. For example if it is now possible to add a dense and a sparse DomainMatrix then there needs to be a test for that. You should also check code coverage of the tests.

I'll add the tests and check for code coverage

ks147 committed:
Added tests for binary operations and added new funcs unify_fmt and unify_domain

ks147 commented:
![image](https://user-images.githubusercontent.com/37146279/112136614-322d2580-8bf5-11eb-88df-0780bd284456.png)

ks147 commented:
@oscarbenjamin, please review

oscarbenjamin commented on sympy/polys/matrices/domainmatrix.py:
```diff
@@ -267,19 +422,14 @@ def unify(self, other):
+        if domain:
+            return self._unify_domain(other)
+        else:
+            return self._unify_fmt(other, fmt)
```
This should unify both the domain and the format rather than one or the other.

oscarbenjamin commented on sympy/polys/matrices/domainmatrix.py:
```diff
@@ -444,8 +596,13 @@ def add(A, B):
             raise ValueError("domain")
+        if isinstance(A.rep, list) != isinstance(B.rep, list):
+            A = A.to_dense()
+            B = B.to_dense()
```
The point of adding the dense/sparse unification logic in `unify` is so that we don't need to do it in each other method.

ks147 replied:
I'll remove it from other methods

ks147 commented on sympy/polys/matrices/ddm.py:
```diff
@@ -198,6 +211,26 @@ def matmul(a, b):
+                    Anew[i] = Ai = []
+                for Bkij in (Bki):
+                    Ai.append(Bkij)
+            cols += Bkcols
```
Ok, I'll implement it simply. I implemented it this way because that's how it was done in SDM class

ks147 committed:
Made changes suggested in code review

- unify now unifies both domains and format
- Binary functions raise ValueError("format") if DomainMatrix formats don't match
- Simplified logic of hstack in DDM class

ks147 committed:
fixed variable to pass code quality test

oscarbenjamin commented on sympy/polys/matrices/ddm.py:
```diff
@@ -211,23 +211,19 @@ def matmul(a, b):
         return c
 
-    def hstack(A, *B):
+    def hstack(A, B):
```
This doesn't match the signature of `SDM.hstack`

oscarbenjamin commented on sympy/polys/matrices/domainmatrix.py:
```diff
@@ -236,20 +310,98 @@ def to_field(self):
+
+        """
+        if isinstance(self.rep, dict):
+            return self
```
This should test for `SDM` rather than `dict`.

oscarbenjamin commented on sympy/polys/matrices/domainmatrix.py:
```diff
@@ -444,8 +600,12 @@ def add(A, B):
         if A.domain != B.domain:
             raise ValueError("domain")
+        if isinstance(A.rep, list) != isinstance(B.rep, list):
+            raise ValueError("format")
```
This could get repetitive so maybe add a `_check` method like `DDM` has.

ks147 replied:
Ok, I'll add it

oscarbenjamin committed:
feat(polys): Allow DomainMatrix to to be dense or sparse

ks147 commented:
@oscarbenjamin, looks like you fixed most of the things suggested in the code review. Is there anything left to change?

oscarbenjamin committed:
feat(polys): Allow DomainMatrix to to be dense or sparse

Increase test coverage

oscarbenjamin committed:
feat(polys): Allow DomainMatrix to to be dense or sparse

Move string formatting to rep

oscarbenjamin committed:
perf(polys): Speed up multiplication for sparse DomainMatrix

oscarbenjamin committed:
perf(polys): Speed up multiplication for sparse DomainMatrix

oscarbenjamin committed:
feat(polys): Allow DomainMatrix to to be dense or sparse

Fix str-printing of SDM/DDM

oscarbenjamin commented:
Okay, this now has full test coverage and is looking good. I also changed it so that sparse matrices print in a sparse format.

Perhaps one addition following up from this that would be useful is if `from_Matrix` would automatically decide whether to use the sparse or dense representation.

ks147 commented:
> Perhaps one addition following up from this that would be useful is if `from_Matrix` would automatically decide whether to use the sparse or dense representation.

I'll start working on that.
